File: Kalorie/Kalorie/Core/NutritionLabelRecognition/NutritionLabelReading.swift
//
//  NutritionLabelReading.swift
//  Kalorie
//
//  Created by Josef Antoni on 13.09.2026.
//

import Foundation

struct NutritionLabelReading: Equatable {

    // MARK: - Properties

    var scannedCode: String?
    var name: String?
    var weightOfProduct: Double?
    var energyKJ: Double?
    var caloriesPerHundredGrams: Double?
    var fat: Double?
    var fatSaturated: Double?
    var fatUnsaturatedFattyAcids: Double?
    var carbohydrate: Double?
    var carbohydratePureSugar: Double?
    var fiber: Double?
    var protein: Double?
    var salt: Double?
    var portions: [FoodPortionDomain]?

    // MARK: - Init

    init(
        scannedCode: String? = nil,
        name: String? = nil,
        weightOfProduct: Double? = nil,
        energyKJ: Double? = nil,
        caloriesPerHundredGrams: Double? = nil,
        fat: Double? = nil,
        fatSaturated: Double? = nil,
        fatUnsaturatedFattyAcids: Double? = nil,
        carbohydrate: Double? = nil,
        carbohydratePureSugar: Double? = nil,
        fiber: Double? = nil,
        protein: Double? = nil,
        salt: Double? = nil,
        portions: [FoodPortionDomain]? = nil
    ) {
        self.scannedCode = scannedCode
        self.name = name
        self.weightOfProduct = weightOfProduct
        self.energyKJ = energyKJ
        self.caloriesPerHundredGrams = caloriesPerHundredGrams
        self.fat = fat
        self.fatSaturated = fatSaturated
        self.fatUnsaturatedFattyAcids = fatUnsaturatedFattyAcids
        self.carbohydrate = carbohydrate
        self.carbohydratePureSugar = carbohydratePureSugar
        self.fiber = fiber
        self.protein = protein
        self.salt = salt
        self.portions = portions
    }
}

extension NutritionLabelReading {

    // MARK: - Properties

    var recognizedFields: Set<FoodItemFormField> {
        var fields: Set<FoodItemFormField> = []
        if name != nil { fields.insert(.name) }
        if weightOfProduct != nil { fields.insert(.weight) }
        if energyKJ != nil { fields.insert(.energyKJ) }
        if caloriesPerHundredGrams != nil { fields.insert(.calories) }
        if fat != nil { fields.insert(.fat) }
        if fatSaturated != nil { fields.insert(.fatSaturated) }
        if fatUnsaturatedFattyAcids != nil { fields.insert(.fatUnsaturated) }
        if carbohydrate != nil { fields.insert(.carbohydrate) }
        if carbohydratePureSugar != nil { fields.insert(.carbohydrateSugar) }
        if fiber != nil { fields.insert(.fiber) }
        if protein != nil { fields.insert(.protein) }
        if salt != nil { fields.insert(.salt) }
        return fields
    }

    var isEmpty: Bool {
        recognizedFields.isEmpty && scannedCode == nil && (portions?.isEmpty ?? true)
    }

    var isCompleteForAutoCapture: Bool {
        caloriesPerHundredGrams != nil && fat != nil && carbohydrate != nil && protein != nil
    }
}

File: Android/app/src/main/java/antoni/kalorie/core/nutritionlabelrecognition/NutritionLabelReading.kt
package antoni.kalorie.core.nutritionlabelrecognition

import antoni.kalorie.components.FoodItemFormField
import antoni.kalorie.core.models.FoodMeasure
import antoni.kalorie.core.models.FoodPortionDomain

data class NutritionLabelReading(
    val scannedCode: String? = null,
    val name: String? = null,
    val weightOfProduct: Double? = null,
    val energyKJ: Double? = null,
    val caloriesPerHundredGrams: Double? = null,
    val fat: Double? = null,
    val fatSaturated: Double? = null,
    val fatUnsaturatedFattyAcids: Double? = null,
    val carbohydrate: Double? = null,
    val carbohydratePureSugar: Double? = null,
    val fiber: Double? = null,
    val protein: Double? = null,
    val salt: Double? = null,
    val portions: List<FoodPortionDomain>? = null,
    val measure: FoodMeasure? = null,
) {

    // MARK: - Properties

    val recognizedFields: Set<FoodItemFormField>
        get() = buildSet {
            if (name != null) add(FoodItemFormField.NAME)
            if (weightOfProduct != null) add(FoodItemFormField.WEIGHT)
            if (energyKJ != null) add(FoodItemFormField.ENERGY_KJ)
            if (caloriesPerHundredGrams != null) add(FoodItemFormField.CALORIES)
            if (fat != null) add(FoodItemFormField.FAT)
            if (fatSaturated != null) add(FoodItemFormField.FAT_SATURATED)
            if (fatUnsaturatedFattyAcids != null) add(FoodItemFormField.FAT_UNSATURATED)
            if (carbohydrate != null) add(FoodItemFormField.CARBOHYDRATE)
            if (carbohydratePureSugar != null) add(FoodItemFormField.CARBOHYDRATE_SUGAR)
            if (fiber != null) add(FoodItemFormField.FIBER)
            if (protein != null) add(FoodItemFormField.PROTEIN)
            if (salt != null) add(FoodItemFormField.SALT)
            if (measure != null) add(FoodItemFormField.MEASURE)
        }

    val isEmpty: Boolean
        get() = recognizedFields.isEmpty() && scannedCode == null && (portions?.isEmpty() ?: true)

    val isCompleteForAutoCapture: Boolean
        get() = caloriesPerHundredGrams != null && fat != null && carbohydrate != null && protein != null
}
